core/example/longrange/ewald.cpp

```diff
@@ -34,63 +36,34 @@ void TEwald::tune( double accuracy, ParticleList particles, double lx,
-    value_type error_estimate;
 
     auto q = Cabana::slice<Charge>( particles );
-    double q_sum;
 
-    const int N_alpha = 200;
-    const int N_k = 2000;
-
-    const int n_max = particles.size();
-
-    // calculate sum of charge squares
-    Kokkos::parallel_reduce( n_max,
-                             KOKKOS_LAMBDA( const int idx, double &q_part ) {
-                                 q_part += q( idx ) * q( idx );
-                             },
-                             q_sum );
-    Kokkos::fence();
-    // Not sure, but think fence required since default memory or exec space
-    // could be CudaUVM
-
-    double r_max = _r_max = std::min( 0.49 * lx, 0.1 * lx + 1.0 );
-    Kokkos::parallel_reduce(
-        "MinLocReduce", N_alpha * N_k,
-        KOKKOS_LAMBDA( const int &idx, value_type &team_errorest ) {
-            int ia = idx % N_alpha;
-            int ik = idx / N_alpha;
-
-            double alpha = (double)ia * 0.05 + 1.0;
-            double k_max = (double)ik * 0.05;
-
-            double delta_Ur = q_sum * sqrt( 0.5 * r_max / ( lx * ly * lz ) ) *
-                              std::pow( alpha * r_max, -2.0 ) *
-                              exp( -alpha * alpha * r_max * r_max );
-
-            double delta_Uk =
-                q_sum * alpha / PI_SQ * std::pow( k_max, -1.5 ) *
-                exp( -std::pow( PI * k_max / ( alpha * lx ), 2 ) );
-
-            double delta = delta_Ur + delta_Uk;
-            Kokkos::pair<double, double> values( alpha, k_max );
-
-            if ( ( delta < team_errorest.val ) && ( delta > 0.8 * accuracy ) )
-            {
-                team_errorest.val = delta;
-                team_errorest.loc = idx;
-            }
-        },
-        reducer_type( error_estimate ) );
-    Kokkos::fence();
-    // Not sure, but think fence required since default memory or exec space
-    // could be CudaUVM
-
-    _alpha = (double)( error_estimate.loc % N_alpha ) * 0.05 + 1.0;
-    _k_max = (double)( error_estimate.loc / N_alpha ) * 0.05;
-
-    _k_max_int[0] = _k_max_int[1] = _k_max_int[2] = std::ceil( _k_max );
+    const int N = particles.size();
+
+    double l = std::max( std::max( lx, ly ), lz );
+
+    // Fincham 1994, Optimisation of the Ewald Sum for Large Systems
+    // only valid for cubic systems (needs adjustement for non-cubic systems)
+    constexpr double EXECUTION_TIME_RATIO_K_R = 2.0;
+    double p = -log( accuracy );
+    _alpha = pow( EXECUTION_TIME_RATIO_K_R, 1.0 / 6.0 ) * sqrt( p / PI ) *
+             pow( N, 1.0 / 6.0 ) / l;
+    _k_max = pow( EXECUTION_TIME_RATIO_K_R, 1.0 / 6.0 ) * sqrt( p / PI ) *
+             pow( N, 1.0 / 6.0 ) / l * 2.0 * PI;
+    _r_max = pow( EXECUTION_TIME_RATIO_K_R, 1.0 / 6.0 ) * sqrt( p / PI ) /
+             pow( N, 1.0 / 6.0 ) * l;
+    _alpha = sqrt( p ) / _r_max;
+    _k_max = 2.0 * sqrt( p ) * _alpha;
+
+    _k_max_int[0] = ceil( _k_max );
+    _k_max_int[1] = ceil( _k_max );
+    _k_max_int[2] = ceil( _k_max );
```

Comment:
As I understand, k_max_int is no longer needed in this new implementation?

Reply:
That is correct, I removed the elements from the class.